iot/mqttsnippets/gateway_send_data_test.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2019 Google LLC. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package mqttsnippets
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2/google"
+	cloudiot "google.golang.org/api/cloudiot/v1"
+)
+
+var privateKeyRSA = os.Getenv("GOLANG_SAMPLES_IOT_PRIV")
+var pubKeyRSA = os.Getenv("GOLANG_SAMPLES_IOT_PUB")
+var region = "us-central1"
+
+// returns a v1 UUID for a resource: e.g. topic, registry, gateway, device
+func createIDForTest(resource string) string {
+	uuid, err := uuid.NewRandom()
+	if err != nil {
+		log.Fatalf("Could not generate uuid: %v", err)
+	}
+	id := fmt.Sprintf("golang-test-%s-%s", resource, uuid.String())
+
+	return id
+}
+
+// getClient returns a client based on the environment variable GOOGLE_APPLICATION_CREDENTIALS
+func getClient() (*cloudiot.Service, error) {
+	// Authorize the client using Application Default Credentials.
+	// See https://g.co/dv/identity/protocols/application-default-credentials
+	ctx := context.Background()
+	httpClient, err := google.DefaultClient(ctx, cloudiot.CloudPlatformScope)
+	if err != nil {
+		return nil, err
+	}
+	client, err := cloudiot.New(httpClient)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// createRegistry creates a IoT Core device registry associated with a PubSub topic
+func createRegistry(w io.Writer, projectID string, region string, registryID string, topicName string) (*cloudiot.DeviceRegistry, error) {
```

Comment:
Are there samples for all of these helpers? Or they're in another package?

Reply:
Could we put them in the same directory/package? I think the cost of having samples in the same directory is lower than copying the helpers. Feel free to disagree. It's OK the manager is `package main` - these could be, too.

Reply:
Potentially worried about scaling long term. If we continue to add samples at the rate of 1 file per sample, this directory could get very big very fast. Willing to entertain the idea though since it decreases duplication, but I think @broady @gguuss might have more ideas.

Reply:
Anecdotally, we have put one snippet per file in PHP; there are a good number of files already but it's not unmaintainable. If we wanted to coalesce the snippets to each feature, that's one way we could group them and reduce the file count.  Longer term, I want to ship e2e demos showing the features, which could be put in the main package.

Reply:
I'd be willing to move to a one directory snippet, but as of right now, the refactor into one directory would cause a lot of issues. Basically it would involve updating the docs to point to a different directory (since I can't just move MQTT samples into the `manager` directory as well as fixing the tests as well. Also, it would make this specific PR very large since there will be a good amount of renaming files and such. I would prefer if we pushed this to a later time and keep it the way it is for now.

Reply:
Also, having a single directory for device management and MQTT samples would be hard for pages like this: https://cloud.google.com/iot/docs/samples/mqtt-samples which have links that point to a specific directory.

Reply:
Got it. Yeah, if the directory is linked from a doc, it's probably good to keep it clean.